### PR TITLE
fp(hufman): Add solution for Hufman coding

### PR DESCRIPTION
**Pull request by Alex-42-AI**
Add hufman coding solution

### CODE REVIEW

File: project/src/Lib.hs [hunk 1]
```diff
@@ -1,4 +1,191 @@
...
+    bits <- readFile "decodeBitsInput.txt"
+    print (decode (makeTreeAndLeaves tree) bits)
+
+in_keys :: Eq a => a -> Dict a b -> Bool
+in_keys _ [] = False
+in_keys x (y:ys) = if x == (fst y) then True else in_keys x ys
+
+
+value :: Eq a => a -> Dict a b -> b
+value _ [] = error "Key not in dictionary!"
```
Comment (googleson78) [inline project/src/Lib.hs]: You should not crash - you should use a `Maybe` here.

File: project/src/Lib.hs [hunk 2]
```diff
@@ -1,4 +1,191 @@
...
+
+main :: IO()
+main = do
+    word <- readFile "encodeInput.txt"
+    print (encode word)
+    tree <- readFile "decodeTreeInput.txt"
+    bits <- readFile "decodeBitsInput.txt"
+    print (decode (makeTreeAndLeaves tree) bits)
+
+in_keys :: Eq a => a -> Dict a b -> Bool
```
Comment (googleson78) [inline project/src/Lib.hs]: You should use a `Maybe` instead of returning a `Bool` and relying on this information.

File: project/src/Lib.hs [hunk 3]
```diff
@@ -1,4 +1,191 @@
...
+
+histogram :: Eq a => [a] -> Dict a Int
+histogram s = helper s []
+  where
+    helper [] res = res
+    helper (c:cs) res = if in_keys c res then helper cs (increment c res) else helper cs ((c, 1):res)
+
+
+least :: Dict a Int -> a
+least [] = error "Empty dictionary!"
```
Comment (googleson78) [inline project/src/Lib.hs]: You should not `error` here, use a `Maybe` instead.

File: project/src/Lib.hs [hunk 4]
```diff
@@ -1,4 +1,191 @@
...
+
+remove :: Eq a => a -> [a] -> [a]
+remove el l = helper el l []
+  where
+    helper _ [] res = reverse res
+    helper x (y:ys) res = if x == y then helper x ys res else helper x ys (y:res)
+
+
+root :: Tree a -> a
+root Null = error "Empty tree!"
```
Comment (googleson78) [inline project/src/Lib.hs]: You should not `error` here, use a `Maybe` instead.

File: project/src/Lib.hs [hunk 5]
```diff
@@ -1,4 +1,191 @@
...
+
+least :: Dict a Int -> a
+least [] = error "Empty dictionary!"
+least (p:ps) = helper (fst p) ps (snd p)
+  where
+    helper el [] _ = el
+    helper el (x:xs) val = if (snd x) < val then helper (fst x) xs (snd x) else helper el xs val
+
+
+remove :: Eq a => a -> [a] -> [a]
```
Comment (googleson78) [inline project/src/Lib.hs]: You can easily use a `filter` here instead

**Comment by googleson78** [comment]
This should not be in this repo, it should be a PR to your own repo.

File: project/src/Lib.hs [hunk 6]
```diff
@@ -1,4 +1,191 @@
...
-main = putStrLn "Hi!"
+type Dict a b = [(a, b)]
+
+data Tree a = Null | Node a (Tree a) (Tree a)
+    deriving (Eq, Show)
+
+makeTree :: String -> Tree Int
+makeTree "" = Null
+makeTree (c0:r) = 
+  if c0 == 'N'
```
Comment (googleson78) [inline project/src/Lib.hs:16]: You should be using an abstract data type for this. What does `'N'` mean, what does `'u'` mean?

File: project/src/Lib.hs [hunk 7]
```diff
@@ -1,4 +1,191 @@
...
+  if c0 == 'N' 
+  then 
+    (if (head r) == 'u' 
+    then Null 
+    else (Node (getValue r 0) 
+    (makeTree (left r)) 
+    (makeTree (right (tail (rest r)) 1)))) 
+  else (makeTree r)
+  where
+    getValue "" _ = error "Invalid string!"
```
Comment (googleson78) [inline project/src/Lib.hs]: You should use `Maybe` or some other way to handle this, not `error`.

File: project/src/Lib.hs (hunk 6 above)
Comment (Alex-42-AI) [reply project/src/Lib.hs:16]: `'N'` means we either have `"Node"` or `"Null"`. If the second character is `'u'`,then that means we have `"Null"`, and we create an empty tree. If it is `"Node"`, then it must followed by a numerical value, which is extrapolated by the `getValue` function. Forgot to mention, the `makeTree` function expects and only works on legitimate input (the string representation of an actual tree).

**Pull request closed without merging**